scripts/mergesat.patch

```diff
@@ -0,0 +1,228 @@
++inline size_t Solver::nUnits() const { return trail_lim.size() < 1 ? trail.size() : trail_lim[0]; }
+ inline const Lit Solver::getUnit(size_t unit_idx) const
+ {
+     assert(unit_idx < nUnits() && "can only access valid unit clauses");
+diff --git a/minisat/core/SolverTypes.h b/minisat/core/SolverTypes.h
+index 4e366ac..3a150dd 100644
+--- a/minisat/core/SolverTypes.h
++++ b/minisat/core/SolverTypes.h
+@@ -189,6 +189,16 @@ class Clause
+         unsigned onQueue : 1;
+         unsigned size : 30;
+     } header;
++#if defined __clang__
++  #pragma clang diagnostic push
++  #pragma clang diagnostic ignored "-Wzero-length-array"
++#elif defined __GNUC__
++  #pragma GCC diagnostic push
++  #pragma GCC diagnostic ignored "-Wpedantic" // no specific ZLA warning in GCC
++#elif defined _MSC_VER
++  #pragma warning(push)
++  #pragma warning(disable:4200)
++#endif
+     union {
+         Lit lit;
+         float act;
+@@ -196,6 +206,13 @@ class Clause
+         uint32_t touched;
+         CRef rel;
+     } data[0];
++#if defined __clang__
++  #pragma clang diagnostic pop
++#elif defined __GNUC__
++  #pragma GCC diagnostic pop
++#elif defined _MSC_VER
++  #pragma warning(pop)
++#endif
+ 
+     friend class ClauseAllocator;
+ 
+diff --git a/minisat/mtl/Vec.h b/minisat/mtl/Vec.h
+index 4abb5db..e9c0d93 100644
+--- a/minisat/mtl/Vec.h
++++ b/minisat/mtl/Vec.h
+@@ -167,8 +167,7 @@ template <class T> void vec<T>::capacity(int min_cap)
+     int add = imax((min_cap - cap + 1) & ~1, ((cap >> 1) + 2) & ~1); // NOTE: grow by approximately 3/2
+     if (add > INT_MAX - cap) throw OutOfMemoryException();
+ 
+-    T *newdata = (T *)::realloc(data, (cap + add) * sizeof(T));
+-    if (newdata == NULL && errno == ENOMEM) throw OutOfMemoryException();
++    T *newdata = (T *)xrealloc(data, (cap + add) * sizeof(T));
+ 
+     cap += add;
+     data = newdata;
+diff --git a/minisat/mtl/XAlloc.h b/minisat/mtl/XAlloc.h
+index 1f43745..50809cb 100644
+--- a/minisat/mtl/XAlloc.h
++++ b/minisat/mtl/XAlloc.h
+@@ -21,7 +21,6 @@ OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWA
+ #ifndef MergeSat_XAlloc_h
+ #define MergeSat_XAlloc_h
+ 
+-#include <errno.h>
+ #include <stdlib.h>
+ 
+ // To not introduce this multiple times, use this header, as it's rather central
+@@ -41,7 +40,7 @@ class OutOfMemoryException
+ static inline void *xrealloc(void *ptr, size_t size)
+ {
+     void *mem = realloc(ptr, size);
+-    if (mem == NULL && errno == ENOMEM) {
++    if (mem == NULL) {
+         throw OutOfMemoryException();
+     } else
+         return mem;
+diff --git a/minisat/parallel/ParSolver.cc b/minisat/parallel/ParSolver.cc
+index ad4c235..b76566b 100644
+--- a/minisat/parallel/ParSolver.cc
++++ b/minisat/parallel/ParSolver.cc
+@@ -44,7 +44,7 @@ ParSolver::ParSolver()
+   : par_reparsed_options(updateOptions())
+   , parsing(false)
+   , verbosity(0)
+-  , use_simplification(true)
++  , use_simplification(false)
```

Comment:
Where does this patch come from and what does it do?

Reply:
This is disabling simplification by default. The alternative would be to support 2 solver types, with and without simplification, as you do for other solvers.

In case of using MergeSat's simplification, another tweak might be required: not allowing to run iterative simplifications with formula growth. That would be achieved by setting this parameter to 'false' by default: https://github.com/conp-solutions/mergesat/blob/3af84ff119ece9dd0ff6b9083022bb4264909ad1/minisat/simp/SimpSolver.cc#L48

Having an interface that allows to configure the SAT backend would be simpler going forward. Fortunately, MergeSat allows to do that by passing the parameters for the solver via an environment variable, e.g.

```
export MINISAT_RUNTIME_ARGS="-no-grow-iter"
```

would have the same effect.

Reply:
My apologies; I wasn't clear.  My question is a bit higher-level.  Why are we patching MergeSat during the build?  With MiniSAT development is largely frozen.  Is this the case with MergeSat?  A number of these seem like pretty basic fixes so they shouldn't be too controversial?  Patching solvers after download is a hack and we should avoid it when possible.

Reply:
MergeSat is under development, so adding features that are useful for consumers of that solver is possible.

However, using the same default configuration of that solver for all use cases does not work well, due to environment restrictions - in MergeSats case that might mostly be the timeout. While parts of this could be addressed by making things more dynamic inside the solver, there is a lot of ground to cover to get there (and benchmarking to do, eventually adding ML models as we did for other solvers already in the past). The short-cut is to adapt the configuration for each use case accordingly. This patch is an attempt to do so for the/a CBMC use case.

I would love to have access to proper CBMC benchmarking, to get good data. Until then, it boils down to extrapolating from past experience.